Exception when launching Preview Commands Editor

If you write code that fail compilation in the Code tab, and try to launch Preview Commands Editor, you'll get PrivilegedActionException with the compilation error. For example, if you write "to setup", you'll get this:

NetLogo is unable to supply you with more details about this error.  Please report the problem
at https://github.com/NetLogo/NetLogo/issues, or to bugs@ccl.northwestern.edu, and paste the
contents of this window into your report.

java.security.PrivilegedActionException: java.security.PrivilegedActionException: Last procedure doesn't end with END at position 0 in 
 at java.security.AccessController.doPrivileged(Native Method)
 at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
 at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
 at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
 at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
 at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
 at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
 at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
 at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.security.PrivilegedActionException: Last procedure doesn't end with END at position 0 in 
 at java.security.AccessController.doPrivileged(Native Method)
 at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
 at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
 at java.awt.EventQueue$4.run(EventQueue.java:731)
 at java.awt.EventQueue$4.run(EventQueue.java:729)
 ... 9 more
Caused by: Last procedure doesn't end with END at position 0 in 
 at org.nlogo.compiler.CompilerExceptionThrowers$.exception(CompilerExceptionThrowers.scala:24)
 at org.nlogo.compiler.StructureParser.parseProcedure(StructureParser.scala:288)
 at org.nlogo.compiler.StructureParser.parse(StructureParser.scala:134)
 at org.nlogo.compiler.CompilerMain$.compile(CompilerMain.scala:25)
 at org.nlogo.compiler.Compiler$.compileProgram(Compiler.scala:28)
 at org.nlogo.headless.HeadlessModelOpener.openFromMap(HeadlessModelOpener.scala:54)
 at org.nlogo.headless.HeadlessWorkspace.openString(HeadlessWorkspace.scala:531)
 at org.nlogo.app.previewcommands.PreviewCommandsDialog.<init>(PreviewCommandsDialog.scala:40)
 at org.nlogo.app.previewcommands.PreviewCommandsEditor.getPreviewCommands(PreviewCommandsEditor.scala:17)
 at org.nlogo.app.ToolsMenu$$anonfun$12.apply$mcV$sp(ToolsMenu.scala:35)
 at org.nlogo.swing.Menu$$anonfun$addMenuItem$2.apply(Menu.scala:9)
 at org.nlogo.swing.Menu$$anonfun$addMenuItem$2.apply(Menu.scala:9)
 at org.nlogo.swing.RichAction$$anon$6.actionPerformed(Implicits.scala:70)
 at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
 at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
 at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
 at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
 at javax.swing.AbstractButton.doClick(AbstractButton.java:376)
 at javax.swing.AbstractButton.doClick(AbstractButton.java:356)
 at javax.swing.plaf.basic.BasicMenuItemUI$Actions.actionPerformed(BasicMenuItemUI.java:802)
 at javax.swing.SwingUtilities.notifyAction(SwingUtilities.java:1663)
 at javax.swing.JComponent.processKeyBinding(JComponent.java:2882)
 at javax.swing.JMenuBar.processBindingForKeyStrokeRecursive(JMenuBar.java:699)
 at javax.swing.JMenuBar.processBindingForKeyStrokeRecursive(JMenuBar.java:706)
 at javax.swing.JMenuBar.processBindingForKeyStrokeRecursive(JMenuBar.java:706)
 at javax.swing.JMenuBar.processKeyBinding(JMenuBar.java:677)
 at javax.swing.KeyboardManager.fireBinding(KeyboardManager.java:307)
 at javax.swing.KeyboardManager.fireKeyboardAction(KeyboardManager.java:293)
 at javax.swing.JComponent.processKeyBindingsForAllComponents(JComponent.java:2974)
 at javax.swing.JComponent.processKeyBindings(JComponent.java:2966)
 at javax.swing.JComponent.processKeyEvent(JComponent.java:2845)
 at java.awt.Component.processEvent(Component.java:6312)
 at java.awt.Container.processEvent(Container.java:2236)
 at java.awt.Component.dispatchEventImpl(Component.java:4891)
 at java.awt.Container.dispatchEventImpl(Container.java:2294)
 at java.awt.Component.dispatchEvent(Component.java:4713)
 at java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1954)
 at java.awt.DefaultKeyboardFocusManager.dispatchKeyEvent(DefaultKeyboardFocusManager.java:806)
 at java.awt.DefaultKeyboardFocusManager.preDispatchKeyEvent(DefaultKeyboardFocusManager.java:1074)
 at java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:945)
 at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:771)
 at java.awt.Component.dispatchEventImpl(Component.java:4762)
 at java.awt.Container.dispatchEventImpl(Container.java:2294)
 at java.awt.Window.dispatchEventImpl(Window.java:2750)
 at java.awt.Component.dispatchEvent(Component.java:4713)
 at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
 at java.awt.EventQueue.access$500(EventQueue.java:97)
 at java.awt.EventQueue$3.run(EventQueue.java:709)
 at java.awt.EventQueue$3.run(EventQueue.java:703)
 ... 14 more

NetLogo 6.0-PREVIEW-12-15
main: org.nlogo.app.AppFrame
thread: AWT-EventQueue-0
OpenJDK Server VM 1.8.0_66-internal (Oracle Corporation; 1.8.0_66-internal-b17)
operating system: Linux 4.2.0-22-generic (i386 processor)
Scala version 2.11.7
JOGL: (3D View not initialized)
OpenGL Graphics: (3D View not initialized)
model: Untitled

07:37:24.829 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
07:37:24.629 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
07:37:24.427 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
07:37:24.305 DirtyEvent (org.nlogo.app.MainCodeTab) AWT-EventQueue-0
07:37:24.259 DirtyEvent (org.nlogo.app.MainCodeTab) AWT-EventQueue-0
07:37:24.227 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
07:37:24.118 DirtyEvent (org.nlogo.app.MainCodeTab) AWT-EventQueue-0
07:37:24.026 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0
07:37:23.845 DirtyEvent (org.nlogo.app.MainCodeTab) AWT-EventQueue-0
07:37:23.826 PeriodicUpdateEvent (org.nlogo.app.App$$anon$1 (org.nlogo.window.GUIWorkspace)) AWT-EventQueue-0

--- a/src/main/org/nlogo/app/previewcommands/PreviewCommandsDialog.scala
+++ b/src/main/org/nlogo/app/previewcommands/PreviewCommandsDialog.scala
@@ -18,6 +18,7 @@ import org.nlogo.window.EditorColorizer
 import org.nlogo.window.GraphicsPreviewInterface
 import org.nlogo.workspace.Evaluator
 import org.nlogo.workspace.ModelsLibrary.getImagePath
+import org.nlogo.workspace.PreviewCommandsRunner.initWorkspace
 import org.nlogo.workspace.WorkspaceFactory
 
 import javax.swing.AbstractAction
@@ -36,8 +37,7 @@ class PreviewCommandsDialog(
   extends JDialog(owner, title, true) {
   org.nlogo.awt.Fonts.adjustDefaultFont(this)
 
-  private val workspace = workspaceFactory.newInstance
-  workspace.openString(modelContents)
+  private val workspace = initWorkspace(workspaceFactory, _.openString(modelContents))
 
   private var _previewCommands = workspace.previewCommands
   def previewCommands = _previewCommands

--- a/src/main/org/nlogo/workspace/PreviewCommandsRunner.scala
+++ b/src/main/org/nlogo/workspace/PreviewCommandsRunner.scala
@@ -9,6 +9,7 @@ import org.nlogo.api.PreviewCommands
 import org.nlogo.api.SimpleJobOwner
 import org.nlogo.nvm.Workspace
 import org.nlogo.nvm.Procedure
+import org.nlogo.util.Utils.getResourceAsString
 
 import scala.util.Try
 
@@ -40,22 +41,53 @@ object PreviewCommandsRunner {
     this(workspaceFactory, _.open(modelPath))
   }
 
+  def initWorkspace(
+    workspaceFactory: WorkspaceFactory,
+    openModelIn: Workspace => Unit,
+    previewCommands: Option[PreviewCommands] = None): AbstractWorkspaceScala = {
+
+    def newWorkspace(openModelIn: Workspace => Unit) = {
+      val ws = workspaceFactory.newInstance
+      try {
+        // set the seed before opening the model so it affects the `startup` procedure
+        val jobOwner = new SimpleJobOwner(this.getClass.getName, ws.world.mainRNG, classOf[Observer])
+        ws.evaluateCommands(jobOwner, "random-seed 0")
+        openModelIn(ws)
+        ws
+      } catch {
+        case e: Throwable =>
+          // rethrow any exception, but at least try to dispose of the workspace
+          ws.dispose(); throw e
+      }
+    }
+
+    val ws = try newWorkspace(openModelIn) catch {
+      case e: CompilerException =>
+        // If we get a compiler exception when trying to open the model, we
+        // open an empty workspace instead. This way, we get an exception
+        // when trying to compile the preview commands themselves (most likely
+        // something like "Nothing named SETUP has been defined") instead of
+        // getting the error from the main code. This is consistent with the
+        // way widgets behave when there is an error in the code tab.
+        // And in the unlikely case where the preview commands don't depend on
+        // the rest of the model, they'll work. NP 2016-01-12
+        newWorkspace(_.openString(getResourceAsString("/system/empty.nlogo")))
+    }
+    previewCommands.foreach(ws.previewCommands = _)
+    ws
+  }
+
   def apply(
     workspaceFactory: WorkspaceFactory,
     openModelIn: Workspace => Unit,
     previewCommands: Option[PreviewCommands] = None): PreviewCommandsRunner = {
 
-    val ws = workspaceFactory.newInstance
-    val jobOwner = new SimpleJobOwner(this.getClass.getName, ws.world.mainRNG, classOf[Observer])
+    val ws = initWorkspace(workspaceFactory, openModelIn, previewCommands)
 
-    // set the seed before opening the model so it affects the `startup` procedure
-    ws.evaluateCommands(jobOwner, "random-seed 0")
-    openModelIn(ws)
-    previewCommands.foreach(ws.previewCommands = _)
     ws.previewCommands match {
       case compilableCommands: PreviewCommands.Compilable =>
         val procedure = ws.compileCommands(compilableCommands.source)
-        new PreviewCommandsRunner(ws, procedure, jobOwner)
+        new PreviewCommandsRunner(ws, procedure)
       case _ => // non-compilable preview commands
         ws.dispose()
         throw new NonCompilableCommandsException
@@ -66,10 +98,10 @@ object PreviewCommandsRunner {
 
 class PreviewCommandsRunner private (
   workspace: Workspace,
-  procedure: Procedure,
-  jobOwner: JobOwner) {
+  procedure: Procedure) {
 
   lazy val previewImage: Try[BufferedImage] = Try {
+    val jobOwner = new SimpleJobOwner(this.getClass.getName, workspace.world.mainRNG, classOf[Observer])
     try
       workspace.evaluateCommands(jobOwner, "startup", workspace.world.observers, true)
     catch {

